crates/base/src/server.rs

```diff
@@ -188,9 +194,82 @@ pub struct ServerFlags {
     pub graceful_exit_deadline_sec: u64,
 }
 
+#[derive(Debug)]
+pub struct Tls {
+    port: u16,
+    key: PrivateKeyDer<'static>,
+    cert_chain: Vec<CertificateDer<'static>>,
+}
+
+impl Clone for Tls {
+    fn clone(&self) -> Self {
+        Self {
+            port: self.port,
+            key: self.key.clone_key(),
+            cert_chain: self.cert_chain.clone(),
+        }
+    }
+}
+
+impl Tls {
+    pub fn new(port: u16, key: &[u8], cert: &[u8]) -> anyhow::Result<Self> {
+        use rustls_pemfile::read_one_from_slice;
+        use rustls_pemfile::Item;
+
+        let Some((key_item, _)) =
+            read_one_from_slice(key).map_err(|err| anyhow!("can't resolve key: {:?}", err))?
+        else {
+            bail!("invalid key data")
+        };
+
+        let mut cert_chain = vec![];
+        let mut cert_slice = cert;
+        loop {
+            let Some((Item::X509Certificate(cert), remain_cert_slice)) =
+                read_one_from_slice(cert_slice)
```

Comment:
Same here, if the `cert_slice` is fully being passed, why do we need to read in loops?

Reply:
Since we may have a certificate file that contains an intermediate certificate, we need to call `read_one_from_slice` to parse the certificate until the `remain_cert_slice` slice tells us that it is empty.

Of course, there are various helper functions in `rustls_pemfile` that can hide loops, but it's a matter of flavor. 😋

Reply:
I think it's better to be explicit. Let's keep it this way.